Codegen + Structured Kernels Lab Exercise

NivekT opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #62665

This PR is an onboarding exercise and will not be merged to master.

This is an exercise to update the `add` operator to bypass the dispatcher to observe the differences in behavior and performance. As a result of bypassing the dispatcher, many functionalities (e.g. autograd) will be lost and many of the CI tests will fail as expected for the PR.

NivekT committed:
temp exercise commit

exercise commit with test

[ghstack-poisoned]

bdhirsh commented on tools/autograd/derivatives.yaml:
```diff
@@ -192,14 +192,14 @@
+#   other: handle_r_to_c(other.scalar_type(), maybe_multiply(grad, alpha.conj()))
+#   result: self_t + maybe_multiply(other_t, alpha)
+
+# - name: add.Scalar(Tensor self, Scalar other, Scalar alpha=1) -> Tensor
```
You don't actually need to remove the derivative entry for `add.Scalar`, since you're not modifying its behavior anywhere else

NivekT replied:
Changed my code based on this suggestion.

bdhirsh commented on test/test_timing.py:
```diff
@@ -0,0 +1,12 @@
+timer = Timer(
+    stmt="x + y",
+    setup="""
+        x = torch.ones((16,))
```
For benchmarking purposes, you'll get more signal on how much framework overhead you're eliminating if you test with empty tensors, e.g. `torch.ones(0)`.

NivekT replied:
Changed my code based on this suggestion.

bdhirsh commented:
lgtm! We talked in chat about the benchmarking. Some ways to get a higher quality signal:
- Use zero-sized tensors (so we're only measuring framework overhead and not allocation)
- Ensure you build REL_WITH_DEB_INFO=1
- You can also benchmark C++ snippets instead of python snippets to avoid including python interpreter overhead in your benchmarks

NivekT committed:
Update on "Codegen + Structured Kernels Lab Exercise"


This PR is an onboarding exercise and will not be merged to master.

This is an exercise to update the `add` operator to bypass the dispatcher to observe the differences in behavior and performance. As a result of bypassing the dispatcher, many functionalities (e.g. autograd) will be lost and many of the CI tests will fail as expected for the PR.

[ghstack-poisoned]

NivekT commented:
Thanks! I just tried it again with those suggestion and they do reduce the number of instructions (overhead).

NivekT commented:
Closing this because it is done. The CI errors are expected since bypassing the dispatcher for the `add` operator loses many features (e.g. autograd) and will lead to many failures within the test suite.
